subprojects/language-java/src/integTest/groovy/org/gradle/language/java/JavaLanguageIncrementalBuildIntegrationTest.groovy

```diff
@@ -38,4 +38,33 @@ class JavaLanguageIncrementalBuildIntegrationTest extends AbstractJvmLanguageInc
         executedAndNotSkipped ":compileMainJarMainJava", ":createMainJar", ":mainJar"
     }
 
-}
\ No newline at end of file
+    def "task outcome is up to date when no recompilation necessary"() {
+        given:
+        buildFile << '''
+            tasks.withType(org.gradle.api.tasks.compile.AbstractCompile) { task ->
+                task.options.incremental = true
+                task.inputs.file 'input.txt' 
+            }
+        '''.stripIndent()
+        file('input.txt') << 'first run'
+
+        when:
+        succeeds mainCompileTaskName
+
+        then:
+        executedAndNotSkipped mainCompileTaskName
+
+        when:
+        file('input.txt') << 'second run, triggers task execution, but no recompilation is necessary'
+
+        and:
+        succeeds mainCompileTaskName, '--debug'
+
+        then:
+        executed mainCompileTaskName
+        skipped mainCompileTaskName
+
+        and:
+        result.output.contains "Executing actions for task '${mainCompileTaskName}'."
```

Comment:
It could also be `UP-TO-DATE` if up-to-date-checks decided that (e.g. by changing a resource file that's not part of compilation inputs).
But in this case it's the actual task action implementation that decided the task is `UP-TO-DATE`.
I didn't find a better way to assert that than relying on logging output :-/